frontend_csv/src/parser/astMsg.rs

```
@@ -0,0 +1,24 @@
+use crate::scanner::token::CSVTokenType;
+#[derive(Clone, PartialEq, Debug)]
+pub enum AstMsgType {
+    Msgtype,
+    Msgdata,
+    Tlvtype,
+    Tlvdata,
+    None,
+}
+pub trait AstMsg {
+    fn get_msg_type(&self) -> AstMsgType;
+    fn get_values(&self) -> String;
+}
+
+pub fn map_csvtoken(v: Vec<CSVTokenType>) -> AstMsgType {
+    match v[..] {
+        [CSVTokenType::MsgTy, CSVTokenType::LiteralString, CSVTokenType::Number] => {
+            AstMsgType::Msgtype
+        }
+        _ => AstMsgType::None,
+    }
+}
+
+impl AstMsgType {}
```

```suggestion
use crate::scanner::token::CSVTokenType;

#[derive(Clone, PartialEq, Debug)]
pub enum AstMsgType {
    Msgtype,
    Msgdata,
    Tlvtype,
    Tlvdata,
    None,
}

pub trait AstMsg {
    fn get_msg_type(&self) -> AstMsgType;
    fn get_values(&self) -> String;
}

pub fn map_csvtoken(v: Vec<CSVTokenType>) -> AstMsgType {
    match v[..] {
        [CSVTokenType::MsgTy, CSVTokenType::LiteralString, CSVTokenType::Number] => {
            AstMsgType::Msgtype
        }
        _ => AstMsgType::None,
    }
}

impl AstMsgType {}
```

In addition what is the message type `None`

---

the None is a placeholder for vector that has no match

---

we really need it?

---

IMO yes it would make it easier to match... if you have a better way please let me know

---

if this is an invalid state why not return an error?

---

because for example:
if the current buffer is msg -> it would match None  
after adding 4 more token it would become msgtype and it would match -> msgtype token type

---

No, I did not get this! sorry!

In the parser or you are parsing the correct value or you fail! There is no maybe! otherwise it make all messy

---

If you add comments in your function to explain what you are doing for me it is also simple to follow! but for now, this does not make sense to me.

---

> No, I did not get this! sorry!
> 
> In the parser or you are parsing the correct value or you fail! There is no maybe! otherwise it make all messy



> If you add comments in your function to explain what you are doing for me it is also simple to follow! but for now, this does not make sense to me.

I have added more comment hopefully it's understandable

---

I still think that there is a better way to implement the parser, and I'm not sure that the match work on the array btw!

---

Hi I tested it and it worked and also there can be condition.... 
but if you have better idea I am open to it